addons/payment/models/payment_transaction.py

```diff
@@ -36,7 +36,7 @@ def _lang_get(self):
         related='acquirer_id.company_id', store=True, index=True)
     reference = fields.Char(
         string="Reference", help="The internal reference of the transaction", readonly=True,
-        required=True)  # Already has an index from the UNIQUE SQL constraint
+        required=True, index="unique")  # Already has an index from the UNIQUE SQL constraint
```

Comment:
```suggestion
        required=True, index="unique")
```

Reply:
Also, what would be the error message now that it's not explicitly specified?

Reply:
Good point, postgres' raw error messages are really rather bad here:

    ERROR:  duplicate key value violates unique constraint "payment_transaction_reference_key"

is not what anyone wants or expects to see, or what any user will understand in any situation. And if it's caught and replaced by "shit happened" it's not helpful either.

So this probably depends on the mostly-not-actually-currently-existing task of translating pg errors to odoo concepts, and possibly customising them. This mapping is *somewhat* easier than the general case because we can generate (and thus know) constraint names, which makes them easier to map. We can even create a map at load time.